cmd/patchVolume.go

```diff
@@ -0,0 +1,153 @@
+	replicaSet []string
+	size       uint64
+	shared     bool
+	sticky     bool
+}
+
+var (
+	updateReq      *volumeUpdateOpts
+	patchVolumeCmd *cobra.Command
+)
+
+// updateVolumeCmd represents the updateVolume command
+var _ = RegisterCommandVar(func() {
+	updateReq = &volumeUpdateOpts{
+		req: &api.SdkVolumeUpdateRequest{},
+	}
+
+	patchVolumeCmd = &cobra.Command{
+		Use:   "volume [NAME]",
+		Short: "Update field(s) of a portworx volume",
+		Example: `$ px patch  volume test1 --halevel 3
+	This set halevel of volume to 3.`,
+
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return fmt.Errorf("Must supply a name for the volume")
+			}
+			return nil
+		},
+		RunE: updateVolume,
+	}
+})
+
+var _ = RegisterCommandInit(func() {
+	patchCmd.AddCommand(patchVolumeCmd)
+
+	patchVolumeCmd.Flags().Int64Var(&updateReq.halevel, "halevel", 0, "New replication factor (Valid Range: [1, 3]) (default 1)")
+	patchVolumeCmd.Flags().StringSliceVarP(&updateReq.replicaSet, "nodes", "n", []string{}, "")
+	patchVolumeCmd.Flags().Uint64VarP(&updateReq.size, "size", "s", 0, "New size for the volume (GiB) (default 1)")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "shared", "r", false, "Shared is true if this volume can be remotely accessed [default false].")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "sticky", "t", false, "set sticky setting [default false].")
+	patchVolumeCmd.Flags().SortFlags = false
+})
+
+func updateVolume(cmd *cobra.Command, args []string) error {
+	ctx, conn, err := PxConnectDefault()
+	ok := false
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	// fetch the volume name from args
+	updateReq.req.VolumeId = args[0]
+
+	// By default blank replicaset needs to be passed due to know volume driver issuw (pwx-9500)
+	// If user provides one it will be overriden
+	updateReq.req.Spec = &api.VolumeSpecUpdate{
+		ReplicaSet: &api.ReplicaSet{
+			Nodes: updateReq.replicaSet,
+		},
+	}
+
+	// check if halevel providied is valid one
+	if updateReq.halevel > 0 {
+		updateReq.req.Spec.HaLevelOpt = &api.VolumeSpecUpdate_HaLevel{
+			HaLevel: int64(updateReq.halevel),
+		}
+		ok = true
+	}
+
+	// specifying replica set nodes for halevel
+	if len(updateReq.replicaSet) != 0 {
+		updateReq.req.Spec.ReplicaSet = &api.ReplicaSet{
+			Nodes: updateReq.replicaSet,
+		}
+		ok = true
+	}
+
+	// check prvoide size is valid
+	if updateReq.size > 0 {
+		//Provided size has to be converted to bytes
+		updateReq.req.Spec.SizeOpt = &api.VolumeSpecUpdate_Size{
+			Size: (updateReq.size * 1024 * 1024 * 1024),
+		}
+		ok = true
+	}
+
+	// For setting volume as shared or not
+	if updateReq.shared {
+		updateReq.req.Spec.SharedOpt = &api.VolumeSpecUpdate_Shared{
+			Shared: updateReq.shared,
+		}
+		ok = true
+	}
+
+	// for setting volume to be sticky
+	if updateReq.sticky {
+		updateReq.req.Spec.StickyOpt = &api.VolumeSpecUpdate_Sticky{
+			Sticky: updateReq.sticky,
+		}
+		ok = true
+	}
+
+	if !ok {
```

Comment:
Suggestion: maybe use reflect.DeepEqual and compare updateReq.req.Spec with an empty updateReq.req.Spec. If it is the same, then no settings were set.

Reply:
It's not exactly empty struct but with replicaset filled (default value is empty). Sounds good, but we would need a deepcopy library vendored in. Can I use this one ```import "github.com/mohae/deepcopy"```